### PR TITLE
feat(core): options object to supersede bit flags for `inject()`

### LINKED ISSUE #46251: Invalid `inject()` function typing
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

`inject()`'s type description seems to be invalid:
https://github.com/angular/angular/blob/5cfde8b23e72cff4c9359f765a462b474bb32631/packages/core/src/di/injector_compatibility.ts#L78-L80

Return type should be `T | null` only if flag is set to `InjectFlags.Optional` meaning that the signature should be something like this:

```typescript
 export function ɵɵinject<T>(token: ProviderToken<T>): T; 
 export function ɵɵinject<T>(token: ProviderToken<T>, flags?: InjectFlags.Default | InjectFlags...): T; 
 export function ɵɵinject<T>(token: ProviderToken<T>, flags?: InjectFlags.Optional): T | null; 
 export function ɵɵinject<T>(token: ProviderToken<T>, flags = InjectFlags.Default): T|null
```

### Please provide a link to a minimal reproduction of the bug

_No response_

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by alxhub**
`inject()` originated as a private API and was made public to support
`InjectionToken` factories in Ivy. For code-size and performance reasons,
when we code generate `inject()` calls we use a bit field to indicate the
various injection modes (optional, skip-self, etc). However, this doesn't
make for a very nice public API.

This commit introduces an alternative object-based API for options. All 4
flags are supported as `boolean` fields on an options object, and converted
to bit flags internally. If TypeScript can prove that `optional` injection
is not requested, it can narrow the return type and remove the `null` type.

DEPRECATION:

The bit field signature of `inject()` has been deprecated, in favor of the
new options object.

Fixes #46251

### CODE REVIEW

**Review by atscott — APPROVED** [review]
reviewed-for: public-api, fw-core

**Comment by atscott** [comment]
One thing to consider would be to include a migration along with the deprecation (or work on one immediately). We've often deprecated things and then totally forgotten about them / pushed the migration work off. I think we should start including migrations along with the deprecation more often.

File: packages/core/test/acceptance/di_spec.ts [hunk 1]
```diff
@@ -3328,6 +3328,102 @@ describe('di', () => {
...
+        });
+
+        @Component({
+          standalone: true,
+          template: '',
+        })
+        class TestCmp {
+          // TypeScript will check if this assignment is legal, which won't be the case if
+          // inject() erroneously returns a `string|null` type here.
+          value: string = inject(TOKEN, {optional: false});
```
Comment (JoostK) [inline packages/core/test/acceptance/di_spec.ts:3421]: We should probably have a similar assignment but without an options object entirety

File: goldens/public-api/core/index.md [hunk 2]
```diff
@@ -600,9 +600,17 @@ export const Inject: InjectDecorator;
 // @public (undocumented)
 export function inject<T>(token: ProviderToken<T>): T;
 
-// @public (undocumented)
+// @public @deprecated (undocumented)
 export function inject<T>(token: ProviderToken<T>, flags?: InjectFlags): T | null;
 
+// @public (undocumented)
+export function inject<T>(token: ProviderToken<T>, options: InjectOptions & {
```
Comment (AndrewKushnir) [inline goldens/public-api/core/index.md:607]: It doesn't seem like the golden file contains the `InjectOptions` interface. I believe we should export it via `packages/core/src/core.ts` as well to make it public.

File: packages/core/src/di/injector_compatibility.ts [hunk 3]
```diff
@@ -118,8 +147,33 @@ export function inject<T>(token: ProviderToken<T>): T;
...
+ * @param token A token that represents a dependency that should be injected.
+ * @param options Control how injection is executed. Options correspond to injection strategies
+ *     that can be specified with parameter decorators `@Host`, `@Self`, `@SkipSelf`, and
+ *     `@Optional`.
+ * @returns the injected value if operation is successful.
+ * @throws if called outside of a supported context, or if the token is not found.
+ *
+ * @publicApi
+ */
+export function inject<T>(token: ProviderToken<T>, options: InjectOptions&{optional?: false}): T;
```
Comment (AndrewKushnir) [inline packages/core/src/di/injector_compatibility.ts:163]: I'm probably missing something, but should this be `{optional?: true}`? (it'd probably be great to add a comment for that type to explain why it's `{optional?: false}`)

**Review by AndrewKushnir — COMMENTED** [review]
Looks great! 👍

File: packages/core/src/di/injector_compatibility.ts (hunk 3 above)
Comment (alxhub) [reply packages/core/src/di/injector_compatibility.ts:163]: `InjectOptions` by itself is optionally optional 😉 

`InjectOptions&{optional?: false}` is a narrower type which says "if the `optional` field is present, it _must_ be `false`". That is, `InjectOptions&{optional: false}` is _never_ requesting optional injection, and we can narrow the return type of this signature to `T`.

Having an `{optional: true}` signature would be unsafe. The user may pass an object by reference, in which case TypeScript will only see an `{optional?: boolean}` object, which wouldn't match a signature that requires `{optional: true}` but may still result in `null` at runtime.

In other words, we want `|null` in the return type _unless_ TypeScript can *prove* optional injection wasn't requested.

Comment (AndrewKushnir) [reply packages/core/src/di/injector_compatibility.ts:163]: Cool, thanks for the explaining this! 👍

**Review by AndrewKushnir — APPROVED** [review]
👍

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

File: packages/core/src/di/injector_compatibility.ts [hunk 4]
```diff
@@ -102,6 +102,35 @@ Please check that 1) the type for the parameter at index ${
               index} is correct and 2) the correct Angular decorators are defined for this class and its ancestors.`);
 }
 
+/**
+ * Type of the options argument to `inject`.
+ *
+ * @publicApi
+ */
+export interface InjectOptions {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/core/src/di/injector_compatibility.ts:110]: I deprecated `InjectFlags` instead :)

[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/core/src/di/injector_compatibility.ts:110]: We still use them in `Injector.get` and `EnvironmentInjector.get` interfaces. Should be add similar flags there first and deprecate the `InjectFlags` after that?

File: packages/core/test/acceptance/di_spec.ts (hunk 1 above)
Comment (prasopch) [reply packages/core/test/acceptance/di_spec.ts:3421]: Tung

**Review by prasopch — COMMENTED** [review]
500000

File: goldens/public-api/core/index.md (hunk 2 above)
Comment (prasopch) [reply goldens/public-api/core/index.md:607]: > It doesn't seem like the golden file contains the `InjectOptions` interface. I believe we should export it via `packages/core/src/core.ts` as well to make it public.
